sipyco/keepalive.py

```diff
@@ -0,0 +1,105 @@
+import asyncio
+import logging
+import socket
+import sys
+from typing import Optional
+
+logger = logging.getLogger(__name__)
+
+
+def set_keepalive(
+        sock: socket,
+        after_idle: Optional[int] = None,
+        interval: Optional[int] = None,
+        max_fails: Optional[int] = None,
+):
+    """ Turn on keepalive and set options for socket
+
+    Args:
+        sock: The socket to target
+        after_idle: How long (in seconds) that the socket will be idle for
+            before sending the first keep alive.
+        interval: Time (in seconds) between keep alive packets.
+        max_fails: Number of probes to send before the connection fails.
+            Ignored on Windows, normally 5 or 10.
+
+    The defaults for the arguments are taken from the OS level defaults. These
+    can be controlled using sysctl on Linux, or the registry on Windows.
+    Normally it is sufficient to set these correctly only at one end of the
+    connection, the server is most convenient for this.
+
+    On Windows both or neither of after_idle and interval must be supplied and
+    max_fails is ignored. On all other OSs other than Linux all the arguments
+    are ignored.
+    """
+    # This is pretty portable actually, it works on at least Linux, Windows,
+    # some BSDs and Solaris.
+    sock.setsockopt(socket.SOL_SOCKET, socket.SO_KEEPALIVE, 1)
+
+    if sys.platform.startswith("linux"):
```

Comment:
Other code uses ``os.name``.

Reply:
(``artiq/coredevice/comm.py`` does not, but that's a mistake AFAICT)

Reply:
Unfortunately I don't think `os.name` gives enough granularity for this situation. `os.name` differentiates between "posix", "nt" and "java" according to the [docs](https://docs.python.org/3/library/os.html#os.name). Whilst `SOL_SOCKET`/`SO_KEEPALIVE` is defined by posix, the socket options for setting the level are a linux specific extension. Other posix implementation unfortunately call them different things.